File: src/main/java/net/adventuregame/particles/ParticleShader.java
package net.adventuregame.particles;

import com.chidozie.core.shaders.ShaderProgram;
import org.joml.Matrix4f;
import org.joml.Vector2f;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ParticleShader extends ShaderProgram {

	private static final String VERTEX_FILE = "particles/particleVertexShader.vert";
	private static final String FRAGMENT_FILE = "particles/particleFragmentShader.frag";
	private static final Logger log = LoggerFactory.getLogger(ParticleShader.class);

	private int location_numberOfRows;
	private int location_projectionMatrix;

	public ParticleShader() {
		super(VERTEX_FILE, FRAGMENT_FILE);
	}

	@Override
	protected void getAllUniformLocations() {
		location_numberOfRows = getUniformLocation("numberOfRows");
		log.info("numberOfRowsLocation: " + location_numberOfRows);
		location_projectionMatrix = super.getUniformLocation("projectionMatrix");
		log.info("projectionMatrixLocation: " + location_projectionMatrix);
	}

	@Override
	protected void bindAttributes() {
		super.bindAttribute(0, "position");
		super.bindAttribute(1, "modelViewMatrix");
		super.bindAttribute(5, "texOffsets");
		super.bindAttribute(6, "blendFactor");
	}

	protected void loadNumberOfRows(float numberOfRows) {
		loadFloat(location_numberOfRows, numberOfRows);
	}

	protected void loadProjectionMatrix(Matrix4f projectionMatrix) {
		super.loadMatrix(location_projectionMatrix, projectionMatrix);
	}

}

File: src/main/kotlin/net/adventuregame/particles/ParticleShader.kt
package net.adventuregame.particles

import com.adv.core.shaders.ShaderProgram
import org.joml.Matrix4f
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class ParticleShader : ShaderProgram(VERTEX_FILE, FRAGMENT_FILE) {
    private var location_numberOfRows = 0
    private var location_projectionMatrix = 0

    override fun getAllUniformLocations() {
        location_numberOfRows = getUniformLocation("numberOfRows")
        log.info("numberOfRowsLocation: $location_numberOfRows")
        location_projectionMatrix = super.getUniformLocation("projectionMatrix")
        log.info("projectionMatrixLocation: $location_projectionMatrix")
    }

    override fun bindAttributes() {
        super.bindAttribute(0, "position")
        super.bindAttribute(1, "modelViewMatrix")
        super.bindAttribute(5, "texOffsets")
        super.bindAttribute(6, "blendFactor")
    }

    fun loadNumberOfRows(numberOfRows: Float) {
        loadFloat(location_numberOfRows, numberOfRows)
    }

    fun loadProjectionMatrix(projectionMatrix: Matrix4f) {
        super.loadMatrix(location_projectionMatrix, projectionMatrix)
    }

    companion object {
        private const val VERTEX_FILE = "particles/particleVertexShader.vert"
        private const val FRAGMENT_FILE = "particles/particleFragmentShader.frag"
        private val log: Logger = LoggerFactory.getLogger(ParticleShader::class.java)
    }
}
